pinot-core/src/main/java/org/apache/pinot/core/data/manager/offline/OfflineSegmentDataManager.java

```diff
@@ -0,0 +1,116 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.data.manager.offline;
+
+import javax.annotation.Nullable;
+import org.apache.pinot.core.data.manager.OfflineSegmentFetcherAndLoader;
+import org.apache.pinot.core.data.manager.SegmentDataManager;
+import org.apache.pinot.core.indexsegment.immutable.ImmutableSegment;
+import org.apache.pinot.core.segment.index.loader.IndexLoadingConfig;
+
+
+/**
+ * Segment data manager for immutable segment.
+ */
+public class OfflineSegmentDataManager extends SegmentDataManager {
```

Comment:
This class should be merged into `ImmutableSegmentDataManager`

Reply:
Lazy segment fetching felt like a separate concern. Though I suppose the realtime use case can just instantiate it with a null cache manager

Reply:
Lazy segment fetching should apply to all immutable segments, regardless of whether they come from offline table or real-time table

Reply:
@kishoreg and I agreed that applying this to real time tables was going to be a lot more difficult; given the complexity around realtime segment fetching.

I tried merging these classes, and it ended up being pretty ugly. They really are separate concerns.

Reply:
I didn't understand the command about complexity of realtime segment fetching. These (realtime, completed segments) are fetched the same way as offline segments.  The consuming segments are never fetched, only uploaded (and/or retained).

Reply:
They aren't fetched in the same way, it's a different code path with added complexity. For realtime, it tries to fetch segments from peers, then from deep store. There's no peer fetching here.